### ISSUE TITLE
specshow: fmax, tmax etc. options don't affect conventional (time, linear freq) spectrograms

### ISSUE THREAD

**Issue by dpwe** [librosa/librosa#331]
I'm plotting a spectrogram using librosa.display.specshow.  I want to zoom in on the first 5 seconds, and the frequencies up to 2 kHz.  I try `librosa.display.specshow(librosa.logamplitude(S), sr=sr, x_axis='time', y_axis='linear', tmax=5.0, fmax=2000)`.  The display still shows the full 30 s timebase and 11 kHz spectrum.

From the docs: 
* fmin, fmax: Used for Mel and CQT scales
* tmin, tmax: Used for _axis = 'tempo'

"That's just weak songwriting, Petey"

**Comment by dpwe** [comment]
See https://github.com/dpwe/librosa/commit/ec0a5562be90a632ab45aac16cec699914735d94 for proposed fix.  I suspect you won't like calling xlim/ylim from within __axis_*, but see what you think.  Also, I realize I need to extend this to __axis_log().

**Comment by bmcfee** [comment]
> I'm plotting a spectrogram using librosa.display.specshow. I want to zoom in on the first 5 seconds, and the frequencies up to 2 kHz. I try librosa.display.specshow(librosa.logamplitude(S), sr=sr, x_axis='time', y_axis='linear', tmax=5.0, fmax=2000). The display still shows the full 30 s timebase and 11 kHz spectrum.

I think maybe the documentation isn't clear here.

These parameters in specshow are for determining the range of the frequency (or tempo) axis and setting its tick marks.  They are not used to control the axis limits directly.

To do what you're asking, it's better to change the axis limits after drawing the spectrogram image.  This way, you won't need to change any axis labeling code, just call it explicitly.

[This notebook](https://gist.github.com/bmcfee/2b0efdf9c17d17ff066a3c3fbff3078c) illustrates how to do what you're asking.

If you have thoughts about how the purpose of the parameters in specshow could be more clearly communicated, I'm keen to revise.

**Comment by dpwe** [comment]
Well, you're right that I originally misunderstood the purpose of tmax and
fmax.  Which is an easy thing to do if you're looking for tmax and fmax
parameters to zoom in the region you're looking for.

Given the number of additional commands involved in actually zooming in per
your example (and the obscurity of controlling time_ticks and
frequency_ticks), I'd say there was a strong case for adding zooming
parameters of this kind directly into specshow.  I agree it's problematic
to overload the roles of fmax and tmax, but maybe we could have some other
parameters for this purpose.

This is a major weakness compared to the Matlab workflow.  It's very common
(surely not only for me) to want to zoom in to detail within an image; in
Matlab you simply type xlim([range I want in the units I see]), and
autoscaling happens.

**Comment by bmcfee** [comment]
> I'd say there was a strong case for adding zooming
> parameters of this kind directly into specshow.  I agree it's problematic
> to overload the roles of fmax and tmax, but maybe we could have some other
> parameters for this purpose.

Agreed, but I'm not sure what the best way to do this is.

One option is to add a display function `lim` that acts as you describe.  So, something like:
```python
>>> librosa.display.specshow(S)
>>> librosa.display.lim([0, 5], axis='x', scale='time', sr=sr, hop_length=hop_length, n_ticks=5, ...)
>>> librosa.display.lim([100, 1000], axis='y', scale='linear', sr=sr, n_fft=n_fft, ...)
```
For conciseness, these could be dispatched by specshow and cut down on the clutter:
```python
>>> librosa.display.specshow(S, x_axis='time', xlim=[0, 5], y_axis='linear', ylim=[100, 1000])
```
this way, you don't have to copypaste the same axis scaling parameters everywhere.  The naming is generally consistent with matplotlib style as well.

I'm not sure how to support some of the weirder ticker formats, eg `chroma` or `tonnetz`.  Maybe we just don't support them, since they have finite(ish) label sets?  Otherwise, the remaining ticker formats can all be inverted to natural coordinates, so something like this ought to be fine:
```python
>>> librosa.display.specshow(C, y_axis='cqt_note', ylim=['C1', 'C3'])
```

Does this all seem reasonable?

**Comment by bmcfee** [comment]
Following up on this:

[PR 368](https://github.com/librosa/librosa/pull/368) substantially overhauls the display submodule.  The biggest change is that spectrogram display (and friends) all operate in natural coordinates, so all the desired functionality discussed in this thread can be accomplished via matplotlib builtins.  For example, to show an STFT up to 5 seconds and limit at 2KHz, you can do

```python
>>> librosa.display.specshow(S, x_axis='time', y_axis='linear')
>>> plt.xlim([0, 5])
>>> plt.ylim([0, 2000])
```
The exact same code works for any frequency-scaled axis (log, cqt, cqt_note, etc).  You can also do beat-synchronous plotting in natural time:
```python
>>> _, beats = librosa.beat.beat_track(y=y, sr=sr)
>>> beats = librosa.util.fix_frames(beats, x_max=S.shape[1])
>>> beat_t = librosa.frames_to_time(beats)
>>> Ssync = librosa.util.sync(S, beats)
>>> librosa.display.specshow(Ssync, x_axis='time', y_axis='linear', x_coords=beat_t)
>>> plt.xlim([0, 5])
```

@dpwe Does this address all your needs from earlier?  If so, would you mind perusing the [PR](https://github.com/librosa/librosa/pull/368) -- especially the revised test output images -- to see if it matches your expectations?

**Comment by dpwe** [comment]
I attempted to review on reviewable.  I had a few nitpicks in the first
file.  The plots look great, only the limited mel range is less desirable,
I think it's OK.  Overall, looks fantastic!  Thanks.

  DAn.

**Comment by bmcfee** [comment]
Resolved by merging #368
